Dataviews: sync from Gutenberg

ramonjd opened the pull request:
## Proposed Changes

This PR updates the `@automattic/dataviews` package with the latest from Gutenberg.

I've also added some commentary around the commands in the SYNC.md doc for others.


See: https://github.com/Automattic/wp-calypso/blob/trunk/packages/dataviews/SYNC.md#updates

## Why are these changes being made?

We regularly sync changes from the upstream repository into this one to keep it up-to-date.

## Testing Instructions

CI should pass. Smoke test Calypso. 

Looks like there are only changes from https://github.com/WordPress/gutenberg/pull/70435

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you tested accessibility for your changes? Ensure the feature remains usable with various user agents (e.g., browsers), interfaces (e.g., keyboard navigation), and assistive technologies (e.g., screen readers) (PCYsg-S3g-p2).
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

ramonjd committed:
Changelog and elaborate on steps in SYNC doc

andrewserong commented on packages/dataviews/SYNC.md:
````diff
@@ -65,26 +65,36 @@ git switch --detach gutenberg/trunk
 ```
 
-3. Prepare a PR to update wp-calypso. This is a regular git merge. It's merging the dataviews-trunk branch into the packages/dataviews folder in this repository.
+`git switch --detach gutenberg/trunk` positioning you at the exact state of the Gutenberg project's trunk branch so you can access the source files you want to sync from. This allows us to grab files from a specific point in another repository's history without creating or modifying local branches.
````
```suggestion
`git switch --detach gutenberg/trunk` positions you at the exact state of the Gutenberg project's trunk branch so you can access the source files you want to sync from. This allows us to grab files from a specific point in another repository's history without creating or modifying local branches.
```

ramonjd replied:
Good spotting!

andrewserong commented on packages/dataviews/SYNC.md:
````diff
@@ -65,26 +65,36 @@ git switch --detach gutenberg/trunk
 ```
 
-4. Push to wp-calypso.
+`git checkout -b update/dataviews-package trunk` checks out a new  branch.
````
```suggestion
`git checkout -b update/dataviews-package trunk` checks out a new branch.
```

ramonjd committed:
Update packages/dataviews/SYNC.md

Co-authored-by: Andrew Serong <14988353+andrewserong@users.noreply.github.com>

ramonjd committed:
Update packages/dataviews/SYNC.md

Co-authored-by: Andrew Serong <14988353+andrewserong@users.noreply.github.com>

andrewserong reviewed (approved):
I smoke tested some areas of Calypso using DataForm and manually tested via Storybook and all looking good to me.

Thanks for the added explainers in `SYNC.md`, too, that's going to help for next time 👍

oandregal committed:
Pin to specific wordpress/dataviews version

ramonjd commented:
> I'm not sure that there is any way that prevents merging via GitHub UI but allows it via CLI. Need to look at this separately 

Thanks for trying @oandregal 

Given the small footprint of this update, but it's just easier to port the changes manually. That is, reinstate https://github.com/Automattic/wp-calypso/pull/104238 and merge that.

andrewserong commented:
> Given the small footprint of this update, maybe it's just easier to port the changes manually. That is, reinstate https://github.com/Automattic/wp-calypso/pull/104238 and merge that.

Oh, my understanding was that now that https://github.com/Automattic/wp-calypso/pull/104265 has been merged, if this PR is rebased, the unit tests should (theoretically) pass now?

ramonjd committed:
Sync dataviews changes from WordPress/gutenberg

- Remove padding around data forms labels
- Update changelog
- Improve form alignment for better UX

ramonjd commented:
Okay, that was fun.  I just discovered, at least this is my conclusion so far, that we can do a simple rebase on trunk (origin wp-calypso) for these types of updates.

Here are some steps that worked for me, if they look good to you @andrewserong and @oandregal I'll add them to the docs later.


### Rebasing Dataviews branches on top of wp-calypso trunk

Sometimes we'll keep our merge branch in sync with wp-calypso trunk. To do this, we need to "move" our work from being based on the dataviews subtree to being based on wp-calypso's main trunk

The packages/dataviews package in wp-calypso is maintained as a Git subtree that syncs from the WordPress/Gutenberg repository. This creates a parallel development line:

```
wp-calypso main trunk:     A---B---C---D---E
                               \
dataviews-trunk (subtree):      X---Y---Z
```

We want a clean, linear history based on wp-calypso's main trunk:

```
wp-calypso trunk:          A---B---C---D---E
                                           \
your-feature-branch:                        F---G---H
```


#### Steps

Applying the "just in case" philosophy, it's a good idea to backup your branch first, e.g., `git branch backup-branch-name`

##### 1. Find the original dataviews merge commit

```bash
git log --oneline --graph -10
# * abc123 Your latest commit
# * def456 Your work commit  
# * ghi789 Merge branch 'dataviews-trunk' into feature-branch  <-- This one!
# * jkl012 Your base commit
```

##### 2. Rebase onto trunk
```bash
git fetch --all
# Take all the commits that come after ghi789 and put them on top of origin/trunk instead.
git rebase --onto origin/trunk ghi789 HEAD
```
##### 3. Update branch
```bash
git checkout <your-branch-name>
```

You'll see something like this in the log. Copy the new HEAD hash.

```bash
Warning: you are leaving 5 commits behind, not connected to
any of your branches:

  d0f9a919b4d Pin to specific wordpress/dataviews version    # ← Copy this hash! It's the new HEAD
  74101e5ca15 Update packages/dataviews/SYNC.md
  630fe9cd82d Update packages/dataviews/SYNC.md
  7eb9cbaf1c4 Changelog and elaborate on steps in SYNC doc
  a43cd1b477b Sync dataviews changes from WordPress/gutenberg
```

Then reset onto the new HEAD

```bash
git reset --hard d0f9a919b4d    # ← Use the copied hash
```
> [!NOTE]
> Alternatively you can get and reset onto the new HEAD like so
> ```bash
> # Get the new head commit hash (the latest commit after rebase)
> NEW_HEAD=$(git log --oneline -1 | cut -d' ' -f1)  
> git reset --hard $NEW_HEAD
> ```


Check the confirmation:

```bash
HEAD is now at d0f9a919b4d Pin to specific wordpress/dataviews version
```

##### 4. Push branch

```bash
git push --force-with-lease origin feature-branch
```

##### 5. Verification

```bash
# Check the branch is based on origin/trunk
git merge-base HEAD origin/trunk  # Should match latest trunk commit

# Verify clean history
git log --oneline --graph -8

# Ensure no missing changes
git status
```

ramonjd commented:
@oandregal @andrewserong 

Sorry, I thought it best to start afresh over here: https://github.com/Automattic/wp-calypso/pull/104283

I tried to rebase onto Calypso then cherry pick the relevant merge commits etc but it ended up erasing the Gutenberg commits in the history.

Burn it all down and begin a new.

This is the latest here: https://github.com/Automattic/wp-calypso/pull/104283

Should be same as before
